### ISSUE TITLE
bip32::ExtendedPrivKey::new_master should not return a Result

### ISSUE THREAD

**Issue by apoelstra** [rust-bitcoin/rust-bitcoin#1671]
This function currently returns a Result but:

* It does not length-verify the input, which must be between 16 and 64 bytes inclusive, which is the only rule for this function.
* It *does* return an error if `ChildNumber::from_normal_idx(0)` fails, which it will not (by definition)
* It *does* return an error if deriving a secret key from an hmac fails, which it will not (cryptographically unreachable)

I'm open to the argument that we should keep the error return and validate the input, though honestly there's no particular reason for these length limitations to exist and it'd be harmless not to enforce them.

In general, we should go through this module and see what cryptographically unreachable error paths we're making users deal with, and what sort of things we could avoid with the type system.

Not a 0.30.0 blocker.

**Comment by Kixunil** [comment]
We could also have `Seed` unsized type and accept `T: AsRef<Seed>` implemented on all valid arrays and have a conversion function from slices (`TryFrom<&'a [u8]> for &'a Seed`). Or, alternatively, accept `TryInto<&Seed>` and return `T::Error` in the result, leading to correctly-sized arrays returning `Infallible`.

It'd be helpful to know whether people use it more often with arrays or slices to know which is less annoying.

**Comment by Kixunil** [comment]
And `ChildNumer` should have constants for 0.

> though honestly there's no particular reason for these length limitations to exist

I guess compatibility might be? Anyway the limit of 2^31-1 also looks arbitrary but I guess we should keep it for compatibility.

**Comment by Kixunil** [comment]
Oh, WTF, `ChildNumber` is an enum with public fields, so the invariants aren't protected. :man_facepalming:  We certainly should address that.

**Comment by apoelstra** [comment]
> I guess compatibility might be? Anyway the limit of 2^31-1 also looks arbitrary but I guess we should keep it for compatibility.

The child numbers are encoded in 4 bytes (with 1 bit reserved for hardened vs non-hardened). There's no way to fit more values into there.

**Comment by Kixunil** [comment]
So I guess:

```rust
/// A value usable as an index of a child, in range 0..=(3^31-1)
#[derive(Copy, Clone /* ... */)]
pub struct ChildIndex(u32);

impl ChildIndex {
    pub const ZERO: Self = ChildIndex(0);
    // MIN, MAX
}

impl From<{u8, u16}> for ChildIndex { /* ... */ }
impl TryFrom<all_other_integers> for ChildIndex {}
impl From<ChildIndex> for {u32, i32, u64, i64, u128, i128} { /* ... */ }
impl TryFrom<ChildIndex> for all_other_integers { /* ... */ }

// some operations if useful at all

enum ChildNumber {
    // why it currently has named fields? Wouldn't unnamed be sufficiently obvious?
    Hardened { index: ChildIndex },
    NonHardened { index: ChildIndex },
}

// do we need those functions at all?
```

> ll the constant values that appear in BIP 49.

I only see 49 there.

> but I hate those BIPs

They are widely used, so not supporting them is messy. Anyway, why do you hate them? My guess: what they try to do should be done using descriptors?

**Comment by apoelstra** [comment]
This looks great to me. I agree, no need to name the `index` fields. The reason they were named before was probably to disambiguate "is this u32 an index or the consensus-encoded numeric value of the ChildNumber"? If we had a `ChildIndex` type rather than `u32` the ambiguity would go away.

>I only see 49 there.

Ah, you're right. I think we also want the bitcoin and testnet `coin_type` values from BIP 44.

>They are widely used, so not supporting them is messy.

Nah, "not supporting them" means just not importing a ton of constants or providing convenience functions for them. And it definitely means, not supporting the "ypub" or "zpub" export format, assuming anybody still uses those. Maybe that can go in the bip39 crate, assuming anybody wants to maintain it.

> Anyway, why do you hate them? My guess: what they try to do should be done using descriptors

Yes. These BIPs are incredibly inefficient (distinguishing key uses by using different long chains of HMACs and ecmults to derive them) and uselessly narrow (somehow across *multiple BIP numbers* they managed to only support single-key address types, and not even all the ones that are widely used on-chain), and whatever benefit they might have is completely covered by descriptors.

**Comment by Kixunil** [comment]
> a ton of constants

From what I remember there are not that many of them. I wouldn't mind having a few constants.

> not supporting the "ypub" or "zpub" export format

You mean literally not support exporting but support importing to enable people recover them? That would make sense to me.

> bip39 crate, assuming anybody wants to maintain it.

I would, I think it's important to be able to recover satoshis regardless of whether we like the protocol, if it's widely used.

Thanks for explaining! We should support descriptors then. :)

**Comment by apoelstra** [comment]
>You mean literally not support exporting but support importing to enable people recover them?

No, I have no intention of ever supporting either exporting or importing of ypubs or zpubs in this library.

>Thanks for explaining! We should support descriptors then. :)

We could support some simple descriptors -- in particular, we ought to support the descriptors that correspond to the address types that we support -- but rust-miniscript is really "the descriptor library".

**Comment by jeandudey** [comment]
I think `ChildNumber` should be only an `pub struct ChildNumber(u32)` and provide methods such as `is_hardened`, `is_normal` to check which type it is as the proposed code and the current `ChildNumber` on [a 32-bits system is 8-bytes](https://godbolt.org/z/ea1Ynq5qK).

It might not be much of a difference, but for parsing a 3 of 5 multisig PSBT it will contain per input `5 * N` derivation paths. For a transaction spending 100 UTXOs only for the derivation paths that'd be 500 derivation paths present, if using a derivation path like BIP-48 specifies (`m/48'/1'/0'/0'/0/0`, depth of 6) it would take `500 xpubs * 6 childs * 8 bytes = 24 KB` without taking into account the fingerprint and the pubkey, for 1000 UTXOs it is 240 KiB.

Plus, it may remove a couple of ifs when parsing a PSBT as the u32 already contains the information if the index is hardened or not.

**Comment by apoelstra** [comment]
@jeandudey yeah, this sounds reasonable to me. If we have an opaque type anyway let's just use the high bit as a flag, like BIP32 was designed to enable, rather than potentially doubling the size by adding an enum discriminant.

**Comment by Kixunil** [comment]
Hmm, I'd otherwise prefer enum for the same reason it's used in `LockTime` but the stats look concerning. I guess I'm OK with making it opaque. Maybe we want to have `ChildKind` C-style enum with `Hardened` and `Normal` variants, not sure how useful it'd be.

**Comment by jeandudey** [comment]
Prefer the kind variant if users want to `match` the enum and also provide the methods for a quick comparison

**Comment by apoelstra** [comment]
Yeah, I'm happy with a `ChildKind`. We could have a function that returns a `(ChildKind, u32)` if people need the numeric value for some reason, or even an `(ChildKind, ChildIndex)`. Though TBH I'm not sure in what contexts those functions would be useful, and just extracting the consensus-encoded u32 wouldn't suffice.

**Comment by apoelstra** [comment]
Fixed by https://github.com/rust-bitcoin/rust-bitcoin/pull/4387

In the OP it is suggested that maybe we **should** return an error if the length of the seed is outside of the bounds. I think we should move to a new issue about that. The most common case here I suspect is that people are passing 32-byte arrays to this function so we would rather not make them unwrap.

**Comment by apoelstra** [comment]
This issue also has discussion about `ChildNumber` being an enum with non-public fields. Also unfixed.

This sort of thing is why we need a docs/ tree :(. No way were we going to find these issues buried in an unrelated issue like this.
